python/amici/sbml_import.py

```diff
@@ -1468,48 +1470,46 @@ def _add_conservation_for_non_constant_species(
             stoichiometric_list, *self.stoichiometric_matrix.shape,
             rng_seed=32)
 
+        # sparsify conserved quantities
+        # construct and reduce A * x0 = total_cl
+        A = sp.zeros(len(cls_coefficients), len(ode_model._states))
+        for i_cl, (cl, coefficients) in enumerate(zip(cls_state_idxs,
+                                                      cls_coefficients)):
+            for i, c in zip(cl, coefficients):
+                A[i_cl, i] = sp.Rational(c)
+        rref, pivots = A.rref()
```

Comment:
Brilliant!

Maybe add a comment that this is reduced row echelon form and why this makes sense?

I am also wondering whether we still need the whole monte carlo mumbo jumbo?

Reply:
Thanks. Will do. My feeling is also that we could skip at least certain parts of the algorithm, although I am not sure what exactly. This should be rewritten cleanly at some point. Would leave it for now, though.